core/src/main/java12/com/linecorp/armeria/internal/common/Java12VersionSpecific.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+import com.linecorp.armeria.common.RequestContext;
+
+/**
+ * Implementation of {@link JavaVersionSpecific} using Java 12 APIs.
+ */
+class Java12VersionSpecific extends JavaVersionSpecific {
+
+    @Override
+    String name() {
+        return "Java 12+";
+    }
+
+    @Override
+    public long currentTimeMicros() {
+        final Instant now = Clock.systemUTC().instant();
+        return TimeUnit.SECONDS.toMicros(now.getEpochSecond()) + TimeUnit.NANOSECONDS.toMicros(
+                now.getNano());
```

Comment:
How about extracting this logic to `JavaVersionSpecific.java9currentTimeMicros()` and calls it from here and `Java9VersionSpecific`?

Reply:
> How about extracting this logic to `JavaVersionSpecific.java9currentTimeMicros()` and calls it from here and `Java9VersionSpecific`?

I thought about this but finally gave up, because `java9` and `java12` are conditionally compiled and packaged according to JDK version, that said, when JDK version is 12 the `java9` codes won't be included, that's why we need to replicate the codes in `java9` to `java12` as well.

If we want to enhance the gradle script to include all `javaN` where `N<=CurrentJDK`, then we can extract those into `JavaVersionSpecific.java9currentTimeMicros()` and calls it from here, but this makes it impossible to deal with the situation when we want to work around a bug only in JDK 12 but not in JDK 9, which one do you guys prefer, keep it as is or enhance the gradle script?

Reply:
What I meant was putting the method in `JavaVersionSpecific`.
```java
public class JavaVersionSpecific {

    ...

    protected java9CurrentTimeMicros() {
        final Instant now = Clock.systemUTC().instant();
        return TimeUnit.SECONDS.toMicros(now.getEpochSecond()) + TimeUnit.NANOSECONDS.toMicros(
                now.getNano());
    }

    ...
}

public class Java9VersionSpecific {

    ...

    @Override
    public long currentTimeMicros() {
        return super.java9CurrentTimeMicros();
    }

    ...
}

public class Java12VersionSpecific {

    ...

    @Override
    public long currentTimeMicros() {
        return super.java9CurrentTimeMicros();
    }

    ...
}
```
If it didn't work, I'm sorry about it. 😅

Reply:
🤣 my bad, misunderstood, let's give it a try

Reply:
Seems good, thanks